### ISSUE TITLE
PHPUnit and Generators 

### ISSUE THREAD

**Issue by live627** [sebastianbergmann/phpunit#2149]
#1174 is still an issue. The offending code seems to be in 4.xx, 5.xx, and master.

The example test in that thread still applies.

**Comment by sebastianbergmann** [comment]
Copied here from #1174:
```php
<?php
class GeneratorTest extends PHPUnit_Framework_TestCase
{
    public static function generator($range)
    {
        foreach (range(0, $range) as $x) {
            yield $x;
        }
    }

    public function testCountingAGeneratorFailsWithPHPWarning()
    {
        $this->assertCount(10, self::generator(10));
    }
}
```
```
$ phpunit GeneratorTest
PHPUnit 5.3.2-7-g3e6974d by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 18 ms, Memory: 4.00Mb

There was 1 error:

1) GeneratorTest::testCountingAGeneratorFailsWithPHPWarning
Exception: Cannot rewind a generator that was already run

/home/sb/GeneratorTest.php:13

FAILURES!
Tests: 1, Assertions: 1, Errors: 1.
```

**Comment by Keanor** [comment]
and problem with mocked methods: http://stackoverflow.com/questions/37118022/phpunit-mock-and-php-generators

**Comment by keradus** [comment]
I encounter this again.
With the following change: https://github.com/keradus/PHP-CS-Fixer/commit/fbbc8c1
the `$this->assertSame('somefile.php', $iterator->current()->getFilename());` fails on HHVM.
If I use `assertSame` instead of `assertCount` everything is working on HHVM.

**Comment by Nessworthy** [comment]
Cause is here:
https://github.com/sebastianbergmann/phpunit/blob/master/src/Framework/Constraint/Count.php#L69
Generators cannot be rewound, but assertCount will try and rewind it to its original state if it can.

Sooo, possible solutions are to either add in a case for generators so the rewind call is skipped or add new assertions which specifically target generators.

Thoughts?

**Comment by sebastianbergmann** [comment]
Closed via 4123b6a1032503f8ed6baa753c61e0ae459a272b.
